File: problem/Maximal_Rectangle/maximalRectangle.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalRectangle = function(matrix) {
    if (matrix.length === 0) return 0;
    const m = matrix.length;
    const n = matrix[0].length;
    const hist = new Array(n).fill(0);
    let ans = 0;

    const getArea = (heights) => {
        let st = [];
        let maxArea = 0;
        for (let i = 0; i <= heights.length; i++) {
            let h = (i === heights.length) ? 0 : heights[i];
            while (st.length > 0 && h < heights[st[st.length - 1]]) {
                let height = heights[st.pop()];
                let width = st.length === 0 ? i : i - st[st.length - 1] - 1;
                maxArea = Math.max(maxArea, height * width);
            }
            st.push(i);
        }
        return maxArea;
    };

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            hist[j] = matrix[i][j] === '1' ? hist[j] + 1 : 0;
        }
        ans = Math.max(ans, getArea(hist));
    }
    return ans;
};

File: problem/Maximal_Rectangle/maximalRectangle.py
class Solution:
    def area(self, heights):
        n = len(heights)
        st = []
        max_area = 0

        for i in range(n + 1):
            h = 0 if i == n else heights[i]
            while st and h < heights[st[-1]]:
                height = heights[st.pop()]
                width = i if not st else i - st[-1] - 1
                max_area = max(max_area, height * width)
            st.append(i)
        return max_area

    def maximalRectangle(self, matrix):
        if not matrix or not matrix[0]:
            return 0
        
        m, n = len(matrix), len(matrix[0])
        hist = [0] * n
        ans = 0

        for i in range(m):
            for j in range(n):
                if matrix[i][j] == '1':
                    hist[j] += 1
                else:
                    hist[j] = 0
            ans = max(ans, self.area(hist))
        return ans